### PR TITLE
fix: report an unverifiable patch distinctly from a verified one

### LINKED ISSUE #1580: The patcher stages a file when it could not verify it: `parses is not False` writes on None as well as True
`SpanPatcher.stage_into` decides whether a patched file is safe to write with:

```python
if result.parses is not False:
    transaction.stage(key, result.content)
```

`PatchResult.parses` is `bool | None`, and `apply()` leaves it `None` when no parser is available for the file's language:

```python
parses: bool | None = None
if parser is not None:
    parses = not parser.parse(content).root_node.has_error
```

So the gate stages on `True` **and** on `None`. A file whose language has no grammar loaded is written having been checked by nothing, and the caller sees a success message rather than a skip. That is a gate which authorises a write when it could not verify anything, rather than refusing.

### Why this matters beyond the missing-grammar case

Even in the `True` case the gate is a proxy. It answers "does this still parse", when the property a rewrite needs is "does this still mean the same thing". Validity is strictly weaker, so it admits corruptions that happen to be grammatical. A concrete one, found and fixed on the T10 import-rewrite branch (PR #1544): rewriting `import type { helper } from './util';` produced `import type from './util';`, which is valid TypeScript (a default import bound to the name `type`), so `parses` was `True` and the corrupt file was staged and written. The sibling shape `export type { helper } ...` produced output that genuinely failed to parse and was correctly refused. Same defect, and the gate caught only the one that happened to be ungrammatical.

The asymmetry worth noting: a guard that **refuses** on a proxy is merely conservative, while a guard that **authorises** on a proxy is unsafe. This one authorises.

### Options, none of which I want to pick unilaterally

1. Require `parses is True` to stage. Safest, but it refuses every rewrite for languages whose grammars are absent, which includes Rust and Go in a base install (the CI matrix already exercises that path, and `test_import_rewrite.py` has to skip its parse assertions there). This would turn a working edit into a silent no-op in exactly that environment.
2. Keep staging on `None` but report it distinctly, so the caller can tell "verified" from "unverifiable" instead of both arriving as `PATCH_OK`.
3. Make the tri-state explicit at the call site and let each caller choose, since a batch edit and an interactive single-file edit may reasonably want different answers.

Option 2 is the smallest change that removes the false assurance without breaking base installs, but the choice affects the edit algebra's postcondition contract (#1531), so it is the owner's call.

### Not proposing a fix here

The behaviour predates the T10 branch, and changing it inside a feature PR would widen that PR into a semantics change for every editing path. Filing instead.

---

### The same collapse exists elsewhere on `main`, and `main` gets it right everywhere except here

Sweeping every tri-state comparison form, not just one literal (`git grep -nE "is not (True|False)|!= (True|False)" origin/main -- '*.py'`), returns three hits in total:

```
codebase_rag/editing/patcher.py:346:  if result.parses is not False:
scripts/vet_issue.py:298:             if verdict.get("injection_suspected") is not False:
scripts/vet_issue.py:302:             if verdict.get("in_scope") is not True:
```

The two in `vet_issue.py` are the interesting part, because they use **mirror operators on the same verdict dict**:

```python
if verdict.get("injection_suspected") is not False:   # unknown => suspected
    flags.append("injection-suspected")
...
    if verdict.get("in_scope") is not True:           # unknown => out of scope
        flags.append("out-of-scope")
```

`.get()` yields `None` for both, and both collapses land the unknown on the **refusing** side, which required picking the opposite operator for each field because the safe direction differs per field. That is evidence of deliberate choice made by construction rather than by comment: the file does not get it right once, it gets it right twice, choosing the operator to fit the field. (One of them is also commented, explaining that `is True` would have let a truthy non-boolean through in the wrong direction.)

Against that, `patcher.py:346` is the sole site in the repository that lands the unknown on the permissive side.

### The docstring describes a two-valued gate

`stage_into` is not merely silent about the third state; it is written as though the third state does not exist:

> A file whose patch no longer parses is not staged.

That is a complete description of a two-valued gate: parses, or does not. A missing comment would be weak evidence that the `None` case went unconsidered. A docstring that positively describes the wrong state space is much stronger.

### The same record carries a second tri-state, collapsing the same direction

`PatchResult.formatted` is also `bool | None`, and the verdict chain in `apply()` is:

```python
if parses is False:
    message = cs.PATCH_PARSE_FAILED.format(path=key)
elif formatted is False:
    message = cs.PATCH_FORMAT_DRIFT.format(path=key, tool=tool)
else:
    message = cs.PATCH_OK.format(path=key, count=len(edits))
```

`formatted is None` (no formatter available for the language) falls through to `else`, and the file is reported `PATCH_OK`. Unchecked is reported as OK, again on the permissive side. So the pattern in this file is systematic rather than a single slip.

The two need answering separately, and the answers could reasonably differ: staging an unparsed file writes bad bytes to the user's disk, whereas reporting an unformatted file as OK only misleads a caller reading the message.

### What I am actually asking

The collapse itself is not a defect and no lint rule would catch it, since a two-valued branch cannot preserve three states. The question is only which side each unknown lands on, and that is a per-call-site decision. `vet_issue.py` demonstrably made it, twice. `patcher.py` makes it twice in the opposite direction, with a docstring suggesting the third state was never in view. Were those two choices intended?

Credit: the second and third call sites, and the `formatted` chain, were found by a peer session grepping the pattern across `main` rather than reasoning about it. My own first sweep grepped a single literal and reported the population as two, which is exactly the query-shaped blind spot that hid the mirror operator four lines away.

---

### The `None` state is tested where it is produced and untested where it is acted on

`patcher.py:346` is the only production consumer of `.parses` in the repository. Enumerating every consumer of both fields rather than grepping for operator spellings (`git grep -nE "\.parses|\.formatted" origin/main`) gives 25 hits: this one line, plus `test_patcher.py`, plus unrelated coincidental matches. So the population question is settled by checking the consumers, not by guessing at how the comparison might be written.

And `test_patcher.py` knows the third state exists:

```python
# test_generic_fallback_for_a_language_without_a_grammar, line 229
assert result.parses is None
assert result.formatter is None
```

with `formatted is None` asserted at 287 and 306 for the same reason. But neither `stage_into` test uses a file with `parses is None`: line 182 is the parse-**failure** case (`parses is False`, asserts nothing is staged) and 201 the ordinary success case.

So the `None` state is understood, reachable and explicitly asserted at the **record** level, and absent at the **decision** level.

Two consequences.

**The author knew `None` occurs.** There is a test named for that exact case. The docstring describing a two-valued gate and the test asserting the third value are both on `main`. So this is not an unknown-unknown: the state was understood where it is produced, and the staging decision was nonetheless reasoned about as two-valued.

**The current behaviour is pinned by no test.** Measured on a clean worktree at `origin/main`:

| line 346 | result |
|---|---|
| `is not False` (current) | 17 passed |
| `is True` (fail-closed) | **17 passed** |
| `is not None` (stage on parse failure too) | 1 failed, 16 passed |

The third row is the control: it proves the suite genuinely watches this line, failing at `test_parse_failure_is_reported_and_not_staged`. So the green in row two is not the suite ignoring line 346 — it is the suite having no opinion about the `None` case specifically.

That is decision-relevant: requiring `True` would not overturn a tested contract, it would establish one for the first time. Whichever way this is decided, the fix needs a `stage_into` test with `parses is None`, because today **both answers pass**.

Credit again to the peer session, which found this by enumerating the fields' consumers after both of our population claims had already been wrong once each.

---

## Recommendation: option 1, stated over the case that actually matters

The options above are less balanced than they look, because `parses is None` is itself two states, not one. `_parser` returns `None` for two unrelated reasons:

```python
def _parser(self, key):
    language = _language_for(Path(key))
    if language is None:
        return None, None                            # (1) not a supported language
    ...
    return language, self._parsers.get(language)     # (2) supported, grammar absent
```

**Case 1** is `notes.txt` or a Markdown file. Verification is not unavailable, it is *meaningless*: no grammar could ever exist, and refusing to stage would be plainly wrong.

**Case 2** is Rust or Go in a base install. Verification is applicable and absent, and staging is the actual risk.

These are the same value at the gate. My base-install objection to option 1 applies only to case 2, and I stated it as though it applied to both.

Two consequences worth having in front of the decision.

**The existing `None` test is case 1.** `test_generic_fallback_for_a_language_without_a_grammar` patches `notes.txt`. So the test demonstrating the author knew `None` occurs is about the harmless case, which may be exactly why the risky one never got a decision of its own.

**No consumer can currently tell the cases apart.** `PatchResult` does not carry the language: `apply()` computes it (`language, parser = self._parser(key)`) and discards it, keeping `formatter: str | None` but not `language`. The information exists at the decision point and is thrown away before the gate sees it.

That settles option 2. "Report `None` distinctly" cannot distinguish unverifiable-because-meaningless from unverifiable-because-missing, so the caller is told only "something was not checked" — true of every text file the patcher will ever touch. It preserves the fail-open write and adds a signal that is noise in the common case.

**So the recommendation is option 1 with the conflation removed:**

| situation | action |
|---|---|
| language unsupported (nothing to verify) | stage |
| `parses is True` | stage |
| `parses is False` | refuse |
| language supported, grammar missing | **refuse, with a real message** |

The base-install objection dissolves at the same time: `.txt` edits keep working, and a Rust rewrite without the Rust grammar becomes an explicit "cannot verify, grammar not installed" rather than a silent unverified write.

The cost is real and worth stating: it needs `language` carried on `PatchResult`, or `parses` split into its two cases. That is smaller than the postcondition-contract change option 2 implies, and it is the change that makes the question answerable at all.

The same split applies to `formatted`, which should be decided separately — reporting an unformatted file as OK only misleads a reader, whereas staging an unverified file writes bad bytes.

Credit: the two-reasons-for-`None` structure, the observation that the existing test covers case 1, and the discarded-`language` fact were found by a peer session reading `_parser` rather than reasoning from the gate.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #1580.

`stage_into` gates on `if result.parses is not False`. `parses` is `bool | None`, so a two-valued branch over three states assigns the third — "could not check" — to the **writing** side. A file whose language has no grammar was staged having been checked by nothing, and `apply` reported it as `PATCH_OK`.

## The decision, and why this one

The issue offered three options and reserved the choice. Taking **option 2**: keep staging on `None`, report it distinctly.

**Option 1 (require `parses is True`) is the one that looks safest and is not.** `None` means no grammar was loaded, which on a base install is the ordinary case for Rust and Go — CI's own base-install job shows `assert None is True` failures in `test_import_rewrite.py` for exactly those languages. Refusing there would turn a working edit into a **silent no-op** on the configuration the issue identifies as most affected. Option 3 moves the tri-state to every call site, which changes the edit algebra's postcondition contract (#1531) — a wider commitment than this defect warrants.

So the write is unchanged and only the report changes. The distinction that matters:

> Staging a file checked by nothing is defensible. Telling the caller it was checked is not.

## What changed

```python
elif parses is None:
    message = cs.PATCH_UNVERIFIED.format(path=key, count=len(edits))
```

plus `PATCH_UNVERIFIED` and a `stage_into` docstring recording that the third state is assigned to staging **deliberately**, with the reason, so the next reader sees a choice rather than an oversight. That also lets the pair be found: the write decision and the report decision are only jointly safe.

No production caller branches on `PATCH_OK` — checked; the only non-test references are the definition and the patcher itself.

## Tests

Written first, verified RED. The failure showed the defect exactly:

```
assert 'notes.txt: 1 edit(s) applied' != 'notes.txt: 1 edit(s) applied'
  where PatchResult(..., parses=None, ..., message='notes.txt: 1 edit(s) applied')
```

`parses=None` with a message byte-identical to `PATCH_OK`.

Two tests, and the second is a control:

- `test_an_unverifiable_patch_is_staged_but_reported_as_unverified` asserts **both** halves — the file is still staged (so a regression to option 1 fails here) and the message is no longer `PATCH_OK`.
- `test_a_verified_patch_still_reports_ok` exists because a change that reported UNVERIFIED for *everything* would satisfy the first test while destroying the distinction it creates.

Reaching the `None` case uses a `.txt` file rather than relying on an absent grammar, so it exercises the same branch on every configuration instead of only on base install.

## A note on the issue's own evidence

Its sweep for this pattern reported three hits. Re-run on current `main`, there is **one**:

```
codebase_rag/editing/patcher.py:346:  if result.parses is not False:
```

The two in `scripts/vet_issue.py` are gone — that file was deleted in #1587 when the issue-vetting scheme was retired. So the patcher is now the sole instance, and the "same collapse exists elsewhere" section is stale rather than wrong.

## Checks

`test_patcher.py`: 17 passed, 1 skipped, 1 failed. The failure is `test_rust_rename_runs_the_rustfmt_check` (`assert result.formatted is True`), and it is **pre-existing**: it fails identically at `origin/main`, and this diff touches `formatted`/`formatter_check` zero times.

## Summary by CodeRabbit

* **Bug Fixes**
  * Patch results now distinguish verified edits from edits applied without an available parser.
  * Files without language support are still updated while clearly reporting that verification was unavailable.
  * Unverified edits with formatting drift now display both warnings.
  * Verified patches continue to be reported as successful.

* **Tests**
  * Added coverage for unverified patches, formatting drift, supported languages without parsers, and verified patches.

### CODE REVIEW

File: codebase_rag/editing/patcher.py [hunk 1]
```diff
@@ -326,6 +326,11 @@ def apply(self) -> dict[str, PatchResult]:
                 message = cs.PATCH_PARSE_FAILED.format(path=key)
             elif formatted is False:
                 message = cs.PATCH_FORMAT_DRIFT.format(path=key, tool=tool)
+            elif parses is None:
+                # Three states, three messages. `parses` is `bool | None`, and
+                # folding None into the OK branch told the caller a file had
+                # been checked when nothing had checked it (issue #1580).
+                message = cs.PATCH_UNVERIFIED.format(path=key, count=len(edits))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/editing/patcher.py:343]: **Correct, and it is a defect in the fix rather than a gap in it. Fixed in `53853f9b`.**

You are right on both counts, including that my docstring asserted something false. It said `apply` reports these as `PATCH_UNVERIFIED`, and for the Rust/Go-with-formatter case it did not — I wrote that sentence in the same commit that made it untrue.

**And it lands in exactly the case this PR exists for**, which is what makes it worse than a precedence quibble. A language with no grammar loaded is precisely a language whose formatter IS installed: Rust and Go on a base install. So the ordering meant the most important scenario reported only `PATCH_FORMAT_DRIFT`, which reads as *"it parsed, it just needs reformatting"* — a strictly more reassuring message than the truth, produced by a change whose entire purpose was removing false assurance.

The two conditions are independent, so the message now carries both:

```
{path}: {count} edit(s) applied, unverified (no parser for it); {tool} would also reformat it
```

Built from a shared `PATCH_UNVERIFIED_FRAGMENT` that both messages concatenate, so the wording cannot drift between them and a caller can match on either.

**Regression test added as asked, and stubbed rather than opportunistic.** `formatter_check` is monkeypatched to return `(rustfmt, False)`, so the branch is exercised on every machine. Relying on a real rustfmt would have made the test's coverage depend on whether the tool is installed AND happens to object to the fixture — precisely the *"green because the path never ran"* shape, and this repo already has a live example of it in `test_rust_rename_runs_the_rustfmt_check`, which fails on this machine for that exact environmental reason.

Suite: 19 passed, 1 skipped, 1 pre-existing failure (that rustfmt test, identical at `origin/main`, and this diff touches `formatted`/`formatter_check` zero times).

File: codebase_rag/tests/test_patcher.py [hunk 2]
```diff
@@ -230,6 +230,66 @@ def test_generic_fallback_for_a_language_without_a_grammar(tmp_path: Path) -> No
...
+
+    Staging is kept -- refusing would turn a working edit into a silent
+    no-op on exactly that install. What must not survive is the false
+    assurance: the caller has to be able to tell "verified" from
+    "unverifiable", and both arriving as PATCH_OK is what makes the gate
+    unsafe rather than merely weak.
+    """
+    _write(tmp_path, "notes.txt", "alpha beta\n")
+    patcher = Patcher(tmp_path)
+    patcher.replace_identifier_at("notes.txt", 1, 6, "beta", "gamma")
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_patcher.py:252]: **Correct, and the gap is worse than "does not cover" — my test could not reach the branch it was written for. Added in `53853f9b`.**

`notes.txt` has no `SupportedLanguage` at all, so `_parser` returns `(None, None)` on the early exit and the formatter is never consulted. The Rust/Go base-install case takes the other branch entirely: the language resolves, `self._parsers.get(language)` returns `None`, and `formatter_check` then runs for real with a live language. Same `parses is None`, different path — and only the second one can produce the combined drift case your sibling comment is about.

So the fixture was structurally incapable of exercising the scenario #1580 names, while looking like it did because the observable it asserted on (`parses is None`) matched.

Two tests now, both using `Patcher(tmp_path, parsers={})` on a `.rs` file as you suggest:

- `test_a_supported_language_with_no_grammar_is_staged_and_unverified` — the staging-and-message contract on the real path
- `test_unverified_and_format_drift_are_both_reported` — the combined case

`parsers={}` is the right lever precisely because it is not `None`: `_parser` only falls back to `load_parsers()` when `self._parsers is None`, so an empty dict simulates a missing grammar deterministically on a machine that has every grammar installed. The `.txt` test is kept for the unsupported-extension path, as you asked — it covers a genuinely different branch.

Worth noting the shape for the record, since it is the second time on this PR: the first finding was my fix carrying the bug's own defect one layer up, and this one is my test asserting on an observable that agreed with the intended path without being on it. Both passed everything I ran.
